timemachine/fe/dummy.py

```diff
@@ -581,3 +612,171 @@ def generate_optimal_dg_ag_pairs(mol, core, bond_idxs):
+    core: list of int
+        core indices
+
+    bond_idxs: list of 2-tuples
+        (i,j) tuples of the bonded terms
+
+    bond_params: list of 2-tuples
+        (k, b) force constant, bond length
+
+    angle_idxs: list of 3-tuples
+        (i,j,k) tuples where the angle is defined between i-j-k
+
+    angle_params: list of 2-tuples
+        (k, a) force constant, angle in radians
+
+    torsion_idxs: list of 4-tuples
+        (i,j,k,l) tuples where the torsion is defined over i-j-k-l
+
+    torsion_idxs: list of 3-tuples
+        (k, p, n) force constant, period, phase
+
+    Returns
+    -------
+    keep_angle_flags, keep_torsion_flags
+        boolean flags indicating which angles and torsions we keep
+
+    """
+
+    assert len(bond_idxs) == len(bond_params)
+    assert len(angle_idxs) == len(angle_params)
+    assert len(torsion_idxs) == len(torsion_params)
+
+    bond_kv = {}
+    for ij, params in zip(bond_idxs, bond_params):
+        bond_kv[ordered_tuple(ij)] = params
+
+    keep_angle_flags = []
+
+    def ixn_is_all_core_or_all_dummy(atom_idxs):
+        all_core = np.all([a in core for a in atom_idxs])
+        all_dummy = np.all([a not in core for a in atom_idxs])
+
+        return all_core or all_dummy
+
+    for (i, j, k), params in zip(angle_idxs, angle_params):
+        b0 = ordered_tuple((i, j))
+        b1 = ordered_tuple((j, k))
+        if (b0 in bond_kv) and (b1 in bond_kv):
+            # bonds are present
+            b0_k, b0_l = bond_kv[b0]
+            b1_k, b1_l = bond_kv[b1]
+            if np.all([b0_k > min_bond_k, b1_k > min_bond_k, b0_l > min_bond_l, b1_l > min_bond_l]):
+                bonds_okay = True
+            else:
+                bonds_okay = False
+        else:
+            # missing bonds
+            bonds_okay = False
+
+        if ixn_is_all_core_or_all_dummy((i, j, k)) or bonds_okay:
+            keep_angle_flags.append(1)
+        else:
+            keep_angle_flags.append(0)
+
+    angle_kv = {}
+    for idx, (ijk, params) in enumerate(zip(angle_idxs, angle_params)):
+        if keep_angle_flags[idx]:
+            angle_kv[ordered_tuple(ijk)] = params
+
+    keep_torsion_flags = []
+    for (i, j, k, l), params in zip(torsion_idxs, torsion_params):
+        b0 = ordered_tuple((i, j))
+        b1 = ordered_tuple((j, k))
+        b2 = ordered_tuple((k, l))
+
+        if (b0 in bond_kv) and (b1 in bond_kv) and (b2 in bond_kv):
+            b0_k, b0_l = bond_kv[b0]
+            b1_k, b1_l = bond_kv[b1]
+            b2_k, b2_l = bond_kv[b2]
+            if np.all(
+                [
+                    b0_k > min_bond_k,
+                    b1_k > min_bond_k,
+                    b2_k > min_bond_k,
+                    b0_l > min_bond_l,
+                    b1_l > min_bond_l,
+                    b2_l > min_bond_l,
+                ]
+            ):
+                bonds_okay = True
+            else:
+                bonds_okay = False
+        else:
+            bonds_okay = False
```

Comment:
style preference -- compression

(e.g. `def bond_okay(b): return (b.k > min_bond_k) and (b.l > min_bond_l)`, `def bonds_okay(bonds): return np.all([bond_okay(b) for b in bonds])`)

Reply:
nit: using Python builtin `all` might be preferable: `all(bond_okay(b) for b in bonds)`

Reply:
I'll clean this up.